### ISSUE TITLE
Detect if the browser supports CTAP2 and adjust the user's options accordingly

### ISSUE THREAD

**Issue by mart-w** [kanidm/kanidm#1744]
**Is your feature request related to a problem? Please describe.**

Currently, Kanidm seems to allow the setup of passkey authentication even if the browser does not support CTAP2. This means that, although the user will be prompted to trigger their authentication device (e.g. tap their Yubikey) and the passkey will be functional, they will not be prompted for additional authentication (e.g. their Yubikey PIN), neither during the setup nor during login. If the user comes back with a browser supporting CTAP2 and tries to log in, they will be prompted for authentication, but it seems like that as long as the browser used to create the passkey didn’t support CTAP2, it will always be possible to authenticate without further authentication, as if the passkey were only used as a second factor.

I tested this with Firefox with `security.webauthn.ctap2` set to `false`, which is currently the default. If users were to setup passwordless login with this setup, this would weaken their account security. Setting `security.webauthn.ctap2` to `true` or using a browser like Chromium during setup fixes the security issue as a passkey created with CTAP2 available seems to not work without it. However, this is not something the average user is likely to be aware of.

**Describe the solution you'd like**

Microsoft, for example, implement their dialogues in such a way that the option of passwordless authentication through WebAuthn isn't even shown if the browser doesn't properly support it. This prevents the user from setting up insecure passkeys for passwordless login. U2F is not affected by this and is always available.

I think that Kanidm should behave the same way. If the browser doesn’t support CTAP2, Kanidm should not allow the user to set up a passkey nor offer to use one for login. Instead, it may provide a notice explaining why this option is not available and how the user might go about fixing fixing the issue.

It should also be considered how to migrate current users of insecure passkeys to proper ones utilising CTAP2 and/or alternative secure authentication methods.

**Describe alternatives you've considered**

The alternative would be to leave it as it is -- users will continue to be able to use CTAP1 for passwordless login, degrading their account's security. As the average user is currently unlikely to notice that their "secure" authentication device is being operated in a less secure way, the user should then at least be made aware of that fact and maybe asked for approval.

**Additional context**

N/A

**Comment by Firstyear** [comment]
Sadly firefox is the only browser here with this issue - they've been asked to implement CTAP2 for more than 5 years now, and it was only late last year they accepted a PR that had this support. It seems from my research that they might finally be taking this seriously though and it will become a default soon.

Given this, I'm not sure it's worth adding a temporary fix in the UI for this as it will be enabled as a default soon and we'd just get to remove it anyway?

@yaleman what do you think?

**Comment by mart-w** [comment]
Wouldn’t that be risky considering how long it might take until _everyone_ has a compatible release of Firefox? Especially in enterprises where they may use extended support releases and only update them periodically, that may take quite a while. It wouldn’t be a big deal if it were something that simply doesn’t work, but this is a security feature that _seems_ like it’s working as designed and only fails in a subtle way most users won’t notice. All the while it creates insecure credentials which will keep working in an insecure way even if users eventually upgrade to a compatible version of Firefox.

**Comment by Firstyear** [comment]
We'd need to have a way to fingerprint firefox to precise versions to ensure they actually support CTAP2 by default - because we won't be allowing people to make CTAP1 only creds because that comes with a lot of baggage. 

So really I'm seeing this as a request for "if firefox is less than version X, then in login / credential update, display a warning about passkeys"

**Comment by mart-w** [comment]
I wonder how Microsoft do it (sorry that I keep referencing them, I don’t mean to endorse them, I just noticed that their implementation handles it this way). They seem to have found a way to actually detect if CTAP2 is available rather than blocking browser versions, because WebAuthn does become available in Firefox when I enable `security.webauthn.ctap2`.

**Comment by mart-w** [comment]
[Might this call do the trick?](https://fidoalliance.org/specs/fido-v2.1-rd-20191217/fido-client-to-authenticator-protocol-v2.1-rd-20191217.html#authenticatorGetInfo) I don’t know enough about web programming to tinker with this myself right now, but this seems like just what you need to find out if CTAP2 is available or not.

**Comment by mart-w** [comment]
I've dug a bit deeper, looking how SimpleWebAuthn implements their check, and discovered that [Firefox has a non-standard function that checks if CTAP2 is enabled](https://hg.mozilla.org/releases/mozilla-beta/rev/ff2c3fd26b15). Apparently, they introduced this [precisely to avoid people trying their luck with browser fingerprinting](https://github.com/w3c/webauthn/issues/1173), lol.

Anyway, I did a couple of quick tests in Firefox (with CTAP2 enabled and disabled) and Chromium and the following JavaScript expression seems to do the trick:

```javascript
typeof window.PublicKeyCredential.isExternalCTAP2SecurityKeySupported === 'function' ? await window.PublicKeyCredential.isExternalCTAP2SecurityKeySupported() : true
```

**Comment by yaleman** [comment]
Reading up on this...

1. ~5 years ago, Microsoft built CTAP2 into Windows 10, and Firefox was a problem then.
2. Firefox implemented an undocumented, non-standard call [isExternalCTAP2SecurityKeySupported in bug 1526023](https://bugzilla.mozilla.org/show_bug.cgi?id=1526023) ([source code](https://github.com/mozilla/gecko-dev/blob/8869a88331f534c072c1c837e55984e2a4a3ff51/dom/webidl/WebAuthentication.webidl#LL23C1-L23C1)) at their request.
3. Which the webauthn WG decided was [good enough](https://github.com/w3c/webauthn/issues/1173) for everyone (in 2019) because Firefox was eventually going to support CTAP2.

To replicate Microsoft's handling of the situation, someone in the Kanidm project would have to:
1. build and test dedicated UI flows which supports n-3 versions of Firefox. (114 is current, looks like they *might* have mostly backported CTAP2 to 112?)
  - during cred creation
  - during login
2. which trusts this undocumented (it's not [here](https://developer.mozilla.org/en-US/docs/Web/API/PublicKeyCredential)), call which is [slated for removal](https://bugzilla.mozilla.org/show_bug.cgi?id=1819713) to identify the problem and warn the user.
3. rip it all out again at *some* point when the unknown portion of our tiny user-base which is running an old version of Firefox finally upgrades.

# tl;dr

At *most* I'd suggest Kanidm checks that endpoint and put a big red warning banner on the credential creation screen going "HEY YOU LOOK TO USE A TERRIBLE OLD BROWSER WHICH HATES USERS AND EXPERIMENTS ON YOU LIKE A TINY BUNNY JUST UPGRADE ALREADY" with a link to this thread, or an FAQ item explaining the situation.

Adding extra UI flows is heavyweight and affects every single user (watch the WASM balloon), not just Firefox (on two specific OSen).

**Comment by Firstyear** [comment]
@micolous This gets pretty close to something we could have to consider in webauth-rs - I'd like your thoughts on how we should action this. 

@yaleman It's not part of https://docs.rs/web-sys/latest/web_sys/ and it appears to be an async function, which also means we'll have significant issues calling from wasm (we have similar issues with isconditionalmediation available, where we can't resolve promises from js methods ...).

**Comment by micolous** [comment]
I think the Firefox story is more complicated:

* They use `mozilla/authenticator-rs` on everything but Android (AFAIK), iOS, and Windows.
* [Firefox on Android does not support FIDO2 at all][0]. The Android WebAuthn API is provided by Google Play Services -- so won't work on AOSP or non-GP devices.
* Firefox on iOS is Safari in a trenchcoat, so does everything Safari does.
* Firefox on Windows supports FIDO2 if the Windows WebAuthn API is available (Windows 10).

The main change coming is that they'll start supporting it on everything _except_ Android (eg: including Windows 7). [It doesn't look like anyone has noticed Android is still broken][3]...

> Currently, Kanidm seems to allow the setup of passkey authentication even if the browser does not support CTAP2. This means that, although the user will be prompted to trigger their authentication device (e.g. tap their Yubikey) and the passkey will be functional...

kanidm should be able to make the credential with `uv=required` and check the `uv` flag is set, and we can verify the attestation of the result:

* if a browser doesn't support CTAP2, then `uv=required` won't work.
* if a browser ignores `uv=required` and tries anyway, then the authenticator sets `uv=0`, and kanidm can reject it
* if a browser _also_ modifies the `uv` flag from the authenticator, then [the attestation object][2] will be invalid, and kanidm can reject it.

I don't think Mozilla are going to be able to easily remove that API, lest that they trigger the same fingerprinting issues they're already having.

As an API, that really should be _everywhere_, because it effectively mandates U2F and CTAP2 support to be able to use WebAuthn.

[0]: https://hg.mozilla.org/releases/mozilla-beta/file/58a336fdc050be3119a9bab4622935a7e0fe35ec/dom/webauthn/PublicKeyCredential.cpp#l139
[1]: https://www.w3.org/TR/webauthn-2/#authenticator-data
[2]: https://www.w3.org/TR/webauthn-2/#sctn-generating-an-attestation-object
[3]: https://bugzilla.mozilla.org/show_bug.cgi?id=1814487#c7

**Comment by Firstyear** [comment]
I've added this to web_sys via https://github.com/rustwasm/wasm-bindgen/pull/3489 so we'll see how that goes. I have already got it working as a POC in webauthn-rs, so provided that wasm-bindgen accept this, we will likely be able to do this.

It is disappointing to have to commit to this work around though :(

**Comment by Firstyear** [comment]
> * [Firefox on Android does not support FIDO2 at all](https://hg.mozilla.org/releases/mozilla-beta/file/58a336fdc050be3119a9bab4622935a7e0fe35ec/dom/webauthn/PublicKeyCredential.cpp#l139). The Android WebAuthn API is provided by Google Play Services -- so won't work on AOSP or non-GP devices.

We'll have no way to detect this for those users, so they'll just have to "have a bad time" sadly.

**Comment by Firstyear** [comment]
There is an example of this on https://webauthn.firstyear.id.au/ now @mart-w can you check you get the warning?

**Comment by mart-w** [comment]
@Firstyear Yes, it shows up if I disable CTAP2. Interestingly, disabling CTAP2 also causes the Yubikey to not blink and tapping it when the authentication popup appears does nothing, so I'm actually being prevented from doing inadvertently something stupid. This is not how it behaved when I experimented in Kanidm. Did you implement any additional logic in your demo to enable this behaviour? If so, I very much like it. Telling the user that some things might not work as expected and then just having insecure setup processes fail might be the next best thing to actually implementing designated UI flows while being less intrusive than big red security alerts.

Unfortunately, I currently only have your demo to go by, as the Kanidm instance I experimented with is out of my control and they do seem to be toying around with the settings at the moment. I'll have to set up my own instance to rule out that it’s something about my setup that has changed.

**Comment by Firstyear** [comment]
> @Firstyear Yes, it shows up if I disable CTAP2. Interestingly, disabling CTAP2 also causes the Yubikey to not blink and tapping it when the authentication popup appears does nothing, so I'm actually being prevented from inadvertently doing something stupid. This is not how it behaved when I experimented in Kanidm. Did you implement any additional logic in your demo to enable this behaviour? If so, I very much like it. Telling the user that some things might not work as expected and then just having insecure setup processes fail might be the next best thing to actually implementing designated UI flows while being less intrusive than big red security alerts.

It's a combination of two things - first the CTAP2 detection as discussed here, and an update to webauthn-rs to move from uv=preferred + credProtect -> uv=required. Since FF supports "no uv" then under preferred it falls back to user presence (UP) only. However when with uv=required, FF can't proceed at all. 

> 
> Unfortunately, I currently only have your demo to go by, as the Kanidm instance I experimented with is out of my control and they do seem to be toying around with the settings at the moment. I'll have to set up my own instance to rule out that it’s something about my setup that has changed.

Given your github profile, I know exactly which instance it is that is being toyed with :) 

Anyway, that's the explanation of what you are seeing. 

For now, web_sys have accepted one of my changes, but isCTAP2 is proving more challenging to have added since it's "not standard" but is a defacto standard by firefox. 

I have to ask though, when I tested this I opened firefox and it supported ctap2 out of the box after a reset of it's settings - did you have to manually disable ctap2? Or was it just disabled for you already?

**Comment by mart-w** [comment]
I only ran into the problem because I was trying out the exact test instance you're thinking of and noticed that passkey authentication doesn't prompt me for my PIN, which is what I would've expected. @mweinelt then pointed out the `security.webauthn.ctap2` setting which was indeed the culprit and was set to false on my laptop. When I went to enable the setting on my workstation, though, I noticed it was already enabled there. Maybe it used to be off by default and Firefox on my laptop kept it that way even after the defaults changed.

I think I did play around with the settings trying to get Webauthn to work back when I got my Yubikeys and Firefox did not yet have proper support. I never got it to work and reverted all the changes, but maybe it somehow marked them as changed and didn't update them when it became the default. Or maybe it’s just because my workstation is running a fresher OS install and thus got setup with the new defaults while my laptop retained the old settings. I said in my initial post that CTAP2 isn’t enabled by default, but in retrospect that’s just what I assumed based on my observations then. It seems like at least newer Firefox installs have it enabled by default. Maybe I can find another older install on some machine somewhere and see if it's enabled there, just to rule out that I broke it myself ages ago and only noticed it now.

**Comment by Firstyear** [comment]
If a value in about:config is in "bold" it means it was set by a person. You can set it to "default" with the small arrow on the right. I think from looking at my ff instance ctap2 is default on now :)

**Comment by mweinelt** [comment]
As I maintain firefox in nixpkgs I have a rather huge number of old profiles. Tried opening the one from 104.0 with Firefox 114.0.1 and ctap2 was enabled. That means they're really enabling it on old profiles, except maybe not if you disabled it manually, but who would do that?

**Comment by Firstyear** [comment]
If you were testing webauthn in an older version, set it to true, tested, then back to false rather than "reset" this would cause a manual-false to be set.

**Comment by mart-w** [comment]
Yes, I think that’s what might have happened, although I _thought_ I looked out for the setting being in bold and it wasn’t. But it was also late at night, so no guarantees :smile:

**Comment by mart-w** [comment]
In that case, this whole issue lost a lot of its relevance. I personally think it still can’t hurt to have an extra safety net in cases like this, especially as a single bad setup can degrade an account's security indefinitely, but obviously that’s for you as project maintainers to decide. Maybe it would make sense to keep uv=required, as that sounds (to me) like a generally good idea and provides a last-minute assertion in case a user has a funky setup? This definitely doesn’t seem to be big enough to require UI modifications anymore, though.

**Comment by Firstyear** [comment]
We were going to upgrade to force uv=required anyway, so that's going to happen very soon regardless :) 

If we can get the isCTAP2 into web-sys it's going to be easier to give those warnings to users though, because ff-esr is still a thing though. But IMO our userbase aren't using ff-esr so it's likely that we could be doing work for nothing.

For now, I think we leave this as is unless there is an influx of ff-esr users.

**Comment by mweinelt** [comment]
Firefox 115 ESR is due in ~2 weeks and 102 ESR is supported for three more months, so I think we can sit that out.

**Comment by Firstyear** [comment]
In that case, I'll just upgrade webauthn-rs and we can consider this resolved.

**Comment by micolous** [comment]
[If I understand Firefox's bug on this correctly](https://bugzilla.mozilla.org/show_bug.cgi?id=1814487), they enabled CTAP2 support a couple of times and had to roll it back (in v112 and v113), and looks like they've done it finally for reals in v114.

In Firefox v114 on macOS (using regular release channel), I can see it's been enabled by default.

Earlier today, I attempted to enroll a YubiKey (which supports CTAP 2.0 and U2F) with no PIN set with another website (which I believe is using `uv=preferred`) and Firefox refused to enroll it, and *had no UI to actually set a PIN*.  I was able to enroll it after I disabled the CTAP 2.0 applet using YubiKey Manager (which is expected behaviour for U2F-only keys)... and I could have also set a PIN with YubiKey Manager.
